File: apps/cli/commands/start.ts
/**
 * Start command - Launch the daemon
 */
import { spawn } from "node:child_process";
import * as path from "node:path";
import { ensureConsoleDir, writePidFile, saveConfig, loadConfig, resolvePortHost, getDaemonStatus, loadEnvFile, getEnvFilePath } from "../daemon-manager.js";
import type { StartOptions } from "../types.js";

interface ServerLaunch {
  cmd: string;
  args: string[];
}

/**
 * Resolve how to launch the server:
 * 1. Compiled install: this `console` binary IS the server (multi-call).
 *    Re-executes itself detached with CONSOLE_SERVE=1.
 * 2. CONSOLE_SERVER_BIN env override (explicit path to a server binary/source)
 * 3. Dev fallback: run apps/server/index.ts via bun from the source tree
 */
function resolveServerLaunch(): ServerLaunch & { serveEnv?: boolean } {
  const execBase = path.basename(process.execPath);
  // Under the bun runtime (dev) execPath is bun/bun-debug/node — not our binary.
  const compiled = !["bun", "bun-debug", "node"].includes(execBase);

  const envBin = process.env.CONSOLE_SERVER_BIN;
  if (envBin && !compiled) return { cmd: envBin, args: [] };

  if (compiled) return { cmd: process.execPath, args: [], serveEnv: true };

  const cliCommandsDir = path.dirname(new URL(import.meta.url).pathname);
  const serverPath = path.resolve(cliCommandsDir, "..", "..", "server", "index.ts");
  return { cmd: "bun", args: [serverPath] };
}

export async function startDaemon(options: StartOptions): Promise<void> {
  const status = await getDaemonStatus();

  if (status.running) {
    console.log(`Daemon is already running (PID: ${status.pid})`);
    console.log(`Server: http://${status.host}:${status.port}`);
    console.log(`Uptime: ${status.uptime}`);
    process.exit(1);
  }

  // Resolve effective port/host: CLI flag > env > saved config > default.
  // Running `console start --port 9090` saves 9090; a later bare
  // `console start` reuses it instead of silently resetting to 3000.
  const saved = await loadConfig();
  const { port, host } = resolvePortHost({ port: options.port, host: options.host }, saved);
  const reusedPort = !options.port && !process.env.PORT;
  const reusedHost = !options.host && !process.env.HOST;
  const config = { port, host, logLevel: "info" };
  await saveConfig(config);
  await ensureConsoleDir();
  if (reusedPort || reusedHost) {
    console.log(`Using saved config from ~/.console/config.json (port ${port}, host ${host}).`);
  }

  // Provider keys (e.g. FIRECRAWL_API_KEY) live in ~/.console/env so every
  // daemon picks them up without shell exports. Explicit process env wins.
  const fileEnv = await loadEnvFile();
  const baseEnv = { ...fileEnv, ...process.env };
  const firecrawlStatus =
    process.env.FIRECRAWL_API_KEY ?? fileEnv.FIRECRAWL_API_KEY
      ? "Firecrawl: API key configured"
      : `Firecrawl: no API key (keyless mode) — run 'console env' to add one`;

  // Resolve how to launch the server (binary install or dev source tree)
  const launch = resolveServerLaunch();

  if (options.daemon) {
    // Start as background daemon
    console.log(`Starting console agent daemon...`);
    console.log(`Port: ${port}`);
    console.log(`Host: ${host}`);
    console.log(firecrawlStatus);

    const env = {
      ...baseEnv,
      PORT: port,
      HOST: host,
      CONSOLE_DAEMON: "true",
    };

    const child = spawn(launch.cmd, [...launch.args], {
      detached: true,
      stdio: "ignore",
      env: launch.serveEnv ? { ...env, CONSOLE_SERVE: "1" } : env,
    });

    child.unref();

    // Wait a moment and check if process is still running
    await new Promise((resolve) => setTimeout(resolve, 1000));

    if (child.pid && child.exitCode === null) {
      await writePidFile(child.pid);
      console.log(`Daemon started successfully (PID: ${child.pid})`);
      console.log(`Server: http://${host}:${port}`);
      console.log(`Logs: ~/.console/logs/daemon.log`);
      console.log(`Run 'console logs' to view logs`);
      console.log(`Run 'console stop' to stop the daemon`);
    } else {
      console.log(`❌ Failed to start daemon`);
      process.exit(1);
    }
  } else {
    // Run in foreground
    console.log(`Starting console agent in foreground...`);
    console.log(`Port: ${port}`);
    console.log(`Host: ${host}`);
    console.log(firecrawlStatus);
    console.log(`Press Ctrl+C to stop`);

    const env = {
      ...baseEnv,
      PORT: port,
      HOST: host,
      CONSOLE_DAEMON: "true",
    };

    const child = spawn(launch.cmd, launch.args, {
      stdio: "inherit",
      env: launch.serveEnv ? { ...env, CONSOLE_SERVE: "1" } : env,
    });
    await new Promise<number>((resolve) => child.on("exit", (code) => resolve(code ?? 0)));
    process.exit(0);
  }
}

File: apps/server-go/internal/cli/commands/start.go
// Start command - launch the daemon. Port of apps/cli/commands/start.ts.
package commands

import (
	"fmt"
	"os"
	"os/exec"
	"time"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/cli/daemon"
)

// StartOptions mirrors the TS StartOptions: port/host save-and-reuse,
// daemon selects background vs foreground, dev pins dev storage.
type StartOptions struct {
	Port   string
	Host   string
	Daemon bool
	Dev    bool
}

// serverLaunch is how to spawn the server process.
type serverLaunch struct {
	cmd       string
	args      []string
	serveSelf bool
}

// resolveServerLaunch mirrors the old TS multi-call behavior: the CLI binary
// re-executes itself as the server (CONSOLE_SERVE=1), with an explicit
// CONSOLE_SERVER_BIN path as an escape hatch.
func resolveServerLaunch() (serverLaunch, error) {
	if envBin := os.Getenv("CONSOLE_SERVER_BIN"); envBin != "" {
		return serverLaunch{cmd: envBin}, nil
	}
	exe, err := os.Executable()
	if err != nil {
		return serverLaunch{}, fmt.Errorf("resolve server binary: %w", err)
	}
	return serverLaunch{cmd: exe, serveSelf: true}, nil
}

// mergedEnv returns the daemon environment: ~/.console/env under explicit
// process env (which always wins), plus the given overrides.
func mergedEnv(overrides map[string]string) []string {
	merged := daemon.LoadEnvFile()
	for _, kv := range os.Environ() {
		for i := 0; i < len(kv); i++ {
			if kv[i] == '=' {
				merged[kv[:i]] = kv[i+1:]
				break
			}
		}
	}
	for k, v := range overrides {
		merged[k] = v
	}
	env := make([]string, 0, len(merged))
	for k, v := range merged {
		env = append(env, k+"="+v)
	}
	return env
}

// StartDaemon launches the server as a background daemon or in the
// foreground, saving port/host so a later bare `start` reuses them.
func StartDaemon(options StartOptions) error {
	// Pin the storage mode explicitly: --dev wins, then an explicit
	// CONSOLE_ENV, otherwise production. Set before any dir resolution so
	// pid/config paths and the spawned server agree.
	if options.Dev {
		_ = os.Setenv("CONSOLE_ENV", "dev")
	} else if os.Getenv("CONSOLE_ENV") == "" {
		_ = os.Setenv("CONSOLE_ENV", "production")
	}

	status := daemon.GetStatus()
	if status.Running {
		fmt.Printf("Daemon is already running (PID: %d)\n", status.Pid)
		fmt.Printf("Server: http://%s:%s\n", status.Host, status.Port)
		fmt.Printf("Uptime: %s\n", status.Uptime)
		os.Exit(1)
	}

	// CLI flag > env > saved config > default. A bare `start` reuses the
	// saved port/host instead of silently resetting to 3000.
	saved := daemon.LoadConfig(daemon.ConsoleDir())
	port, host := daemon.ResolvePortHost(options.Port, options.Host, saved)
	reusedPort := options.Port == "" && os.Getenv("PORT") == ""
	reusedHost := options.Host == "" && os.Getenv("HOST") == ""
	if err := daemon.SaveConfig(daemon.Config{Port: port, Host: host, LogLevel: "info"}); err != nil {
		return fmt.Errorf("save config: %w", err)
	}
	if err := daemon.EnsureConsoleDir(); err != nil {
		return err
	}
	if reusedPort || reusedHost {
		fmt.Printf("Using saved config from ~/.console/config.json (port %s, host %s).\n", port, host)
	}

	fileEnv := daemon.LoadEnvFile()
	firecrawlStatus := "Firecrawl: no API key (keyless mode) — run 'console env' to add one"
	if os.Getenv("FIRECRAWL_API_KEY") != "" || fileEnv["FIRECRAWL_API_KEY"] != "" {
		firecrawlStatus = "Firecrawl: API key configured"
	}

	launch, err := resolveServerLaunch()
	if err != nil {
		return err
	}

	overrides := map[string]string{
		"PORT": port, "HOST": host, "CONSOLE_DAEMON": "true",
	}
	if launch.serveSelf {
		overrides["CONSOLE_SERVE"] = "1"
	}
	env := mergedEnv(overrides)

	if options.Daemon {
		fmt.Println("Starting console agent daemon...")
		fmt.Printf("Mode: %s (storage: %s)\n", daemon.ResolveMode(), daemon.ConsoleDir())
		fmt.Printf("Port: %s\n", port)
		fmt.Printf("Host: %s\n", host)
		fmt.Println(firecrawlStatus)

		logFile, err := os.OpenFile(daemon.LogFilePath(), os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
		if err != nil {
			return fmt.Errorf("open log file: %w", err)
		}
		defer logFile.Close()

		child := exec.Command(launch.cmd, launch.args...)
		child.Env = env
		child.Stdin = nil
		child.Stdout = logFile
		child.Stderr = logFile
		child.SysProcAttr = detachedProcAttr()
		if err := child.Start(); err != nil {
			return fmt.Errorf("spawn daemon: %w", err)
		}
		pid := child.Process.Pid
		// Reap in the background: Wait both collects the exit status and
		// prevents a zombie (a zombie still answers kill(pid, 0), which
		// would fake a "running" daemon). The buffered channel never blocks
		// the reaper; the select below (then the cleanup watcher) drains it.
		exited := make(chan error, 1)
		go func() { exited <- child.Wait() }()

		// Wait a moment and check the process is still running — a server
		// that exits during the grace period is a failed start.
		select {
		case err := <-exited:
			if err != nil {
				fmt.Printf("Daemon exited during startup: %v\n", err)
			}
			fmt.Println("❌ Failed to start daemon")
			os.Exit(1)
		case <-time.After(time.Second):
		}
		// Keep draining so a later daemon exit is reaped (stale pid files
		// are cleared lazily by GetStatus).
		go func() { <-exited }()

		if err := daemon.WritePidFile(pid); err != nil {
			return fmt.Errorf("write pid file: %w", err)
		}
		fmt.Printf("Daemon started successfully (PID: %d)\n", pid)
		fmt.Printf("Server: http://%s:%s\n", host, port)
		fmt.Println("Logs: ~/.console/logs/daemon.log")
		fmt.Println("Run 'console logs' to view logs")
		fmt.Println("Run 'console stop' to stop the daemon")
		return nil
	}

	fmt.Println("Starting console agent in foreground...")
	fmt.Printf("Port: %s\n", port)
	fmt.Printf("Host: %s\n", host)
	fmt.Println(firecrawlStatus)
	fmt.Println("Press Ctrl+C to stop")

	child := exec.Command(launch.cmd, launch.args...)
	child.Env = env
	child.Stdin = os.Stdin
	child.Stdout = os.Stdout
	child.Stderr = os.Stderr
	_ = child.Run()
	os.Exit(0)
	return nil
}
